networkx/algorithms/graph_hashing.py

```diff
@@ -78,17 +103,17 @@ def weisfeiler_lehman_graph_hash(
     Omitting the `edge_attr` option, results in identical hashes.
 
     >>> nx.weisfeiler_lehman_graph_hash(G1)
-    '0db442538bb6dc81d675bd94e6ebb7ca'
+    '3108a936907b3ee3704642f9eeab34fe'
```

Comment:
Yes @rossbar you are correct, and well spotted. I had thought this too while developing but didn't question why my hashes were different too much because the tests are thorough.

I've not investigated but this could be to do with the fact we call the blake2b hash each time rather than calling `.update` method. If we are very concerned we could check the unhashed versions of some subgraphs before and after this change (effectively replacing the `_hash_label` function with string concatenation).

I would agree that this isn't too detrimental if there is a release warning. Is the release notes something I can update from this PR?

Reply:
> I would agree that this isn't too detrimental if there is a release warning. Is the release notes something I can update from this PR?

Yes I think we'd want to have a prominent note in `doc/release/release_dev.rst` about the change, maybe in`Highlights` since that's the top section.

Before we do that though, I just want to see what others think about this. In principle, if we wanted to be *really* safe wrt backwards compatibility, we could wait to merge this for the NX 3.0 release, where breaking compatibility with 2.X wouldn't be as big a concern. IMO this is overkill - this PR contains a nice new feature and I think the backwards incompatible part is unlikely to affect users (the use-case I can think of where this might be an issue is if users are generating graph hashes, saving them to file, then loading them later for comparison). So my two cents: add a prominent note in `release_dev` and put it in for 2.7. Thoughts?

Reply:
Yes, a possible usecase would be change detection on a large database of graphs - on upgrading to 2.7 users might find every graph is flagged as being changed unless they get a warning telling them to recompute all hashes first.

Reply:
I'm a little worried that we don't know why the hashes changed. Did we change the order of operations or something like that which might affect the hash?  What is invariant for a hash of a graph? -- certainly the same graph gives the same hash if we do everything in the same order...

Reply:
Ok, I did a bit of preliminary investigation. I've not found the reason for the hash difference between this and main, _but_ I have found a bug on main which persists in this branch. It comes from the `neighborhood_aggregate` function. This was left unchanged in this PR apart from 
```python
prefix = "" if not edge_attr else G[node][nei][edge_attr]` # old
prefix = "" if edge_attr is None else str(G[node][nei][edge_attr]) # new
```
(which is more robust but not should not cause hash changes to well-behaved inputs) and the fact that it was copied outside the WL-hash function so that both hashing methods could use it.

The issue is that this function (written here simplified without edge attributes at all):
```python
def _neighborhood_aggregate(G, node, node_labels):
    label_list = [node_labels[node]]
    for nbr in G.neighbors(node):
        label_list.append(node_labels[nbr])
    return "".join(sorted(label_list))
```
makes no distinction between the original node and its neighbors, since sorting combines both.

This I'm pretty sure is bad behaviour. The simplest counterexample I could come up with was the following:

```python
G = nx.DiGraph()
G.add_nodes_from([
   (1, {'node_attr': 'A'}),
   (2, {'node_attr': 'B'}),
   (3, {'node_attr': 'C'})
])
G.add_edges_from([
   (1, 2),
   (2, 3),
   (3, 1)
])
```
i.e. G is a directed cyclic triangle. Then under the current implementation 
```python
weisfeiler_lehman_graph_hash(G, node_attr='node_attr')
weisfeiler_lehman_graph_hash(G2.reverse(), node_attr='node_attr')
```
the above hashes are equal at any depth, i.e `G` and `G.reverse()` pass an isomorphism test, even though G has an edge from node type A -> B and its reverse does not.

It's probably possible to come up with 2 non-isomorphic undirected graphs which pass due to this technicality.

I believe using instead
```python
def _neighborhood_aggregate(G, node, node_labels):
    label_list = []
    for nbr in G.neighbors(node):
        label_list.append(node_labels[nbr])
    return str(node_labels[node]).join(sorted(label_list))
```
would catch these cases, but certainly this would cause graph hashes to change across the board, requiring the warning. This should definitely be included though, since we have graphs the algorithm is flagging as isomorphic when it should be able to discern they are not (in general the algorithm does not imply isomorphism, just gives a high likelihood in practice)

FYI I caught this by changing the `_hash_label` function to just return the label as an unhashed string, this way the final graph `hash` for the above graph was (after 1 iteration):
```
(('AB', 1), ('AC', 1), ('BC', 1))
```
which is much easier to debug than a random hash.

I'm still going to try and compare with main to see if any operations are swapped regardless of this bug.

Reply:
I've resolved why the hashes were different - in fact it's a bug I fixed early in development and forgot to mention in the description. I'll make a separate comment thread summarizing.